qtwebengine (5.12.10) build failed on macOS

**Describe the bug**

I found a similar issue here: #76715, `/.../-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.`

but not sure why mine is broken. Is it because I installed nix to a custom location?

Full log:

https://gist.github.com/nasyxx/b119b8962b34103404c878e1a3667514

Error message:

```
Done. Made 8971 targets from 1526 files in 3360ms
/Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.
make[3]: Entering directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/3rdparty/ninja/ninja -j16 -C /private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core/release QtWebEngineCore
ninja: Entering directory `/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core/release'
ninja: error: '../../../../../../../usr/lib/libsandbox.1.dylib', needed by 'obj/components/cdm/renderer/librenderer.a', missing and no known rule to make it
make[3]: *** [Makefile.gn_run:360: run_ninja] Error 1
make[3]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
make[2]: *** [Makefile:85: sub-gn_run-pro-make_first] Error 2
make[2]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
make[1]: *** [Makefile:84: sub-core-make_first] Error 2
make[1]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src'
make: *** [Makefile:52: sub-src-make_first] Error 2
error: builder for '/Users/Nasy/.nix/store/b1cjkiwi6vaxfr0a6nmhhkpai7zyiv3p-qtwebengine-5.12.10.drv' failed with exit code 2
```

**To Reproduce**
Steps to reproduce the behavior:

`nix-shell -p "qt5.qtwebengine"`

**Expected behavior**

`qtwebengine` installed

**Notify maintainers**
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.3.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.4pre20210308_1c0e3e4`
find: ‘/nix/var/nix/profiles/per-user’: No such file or directory
```

My profiles is under ~/.nix/var/nix/profiles.

> And /usr/lib/libsandbox.1.dylib is present on the filesystem and you are building without a sandbox?

I cannot use a sandbox.  Looks like macOS Big Bur cannot build with a sandbox.

And now I have solved the sandbox problem by add `pkgs.darwin.apple_sdk.frameworks.CoreMediaIO` to the `propagatedBuildInputs`.

However, another error happened: `Project ERROR: Unknown module(s) in QT: qml-private quick-private`.  I wonder if I missed something again?

```
...
[175/175] LINK gnacode_writer.oter.oer.o
/Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.
make[3]: Entering directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/buildtools'
make[3]: Nothing to be done for 'first'.
make[3]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/buildtools'
make[2]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/buildtools'
cd core/ && ( test -e Makefile || /Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/bin/qmake -o Makefile /private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core/core.pro PREFIX=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_OUT=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_DEV=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev NIX_OUTPUT_BIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin NIX_OUTPUT_DOC=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev/share/doc/qt-5.12.10 NIX_OUTPUT_QML=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/qml NIX_OUTPUT_PLUGIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/plugins CONFIG+=release ) && make -f Makefile
/Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.
make[2]: Entering directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
( test -e Makefile.core_headers || /Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/bin/qmake -o Makefile.core_headers /private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core/core_headers.pro PREFIX=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_OUT=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_DEV=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev NIX_OUTPUT_BIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin NIX_OUTPUT_DOC=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev/share/doc/qt-5.12.10 NIX_OUTPUT_QML=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/qml NIX_OUTPUT_PLUGIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/plugins CONFIG+=release ) && make -f Makefile.core_headers
/Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.
make[3]: Entering directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
make[3]: Nothing to be done for 'first'.
make[3]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
( test -e Makefile.core_generator || /Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/bin/qmake -o Makefile.core_generator /private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core/core_generator.pro PREFIX=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_OUT=/Users/Nasy/.nix/store/i9hrvx8cpa4ggr2dka82jrvg0qc0dvmf-qtwebengine-5.12.10 NIX_OUTPUT_DEV=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev NIX_OUTPUT_BIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin NIX_OUTPUT_DOC=/Users/Nasy/.nix/store/fklf437bqdgzqyr3rwhfkwl50gidx07f-qtwebengine-5.12.10-dev/share/doc/qt-5.12.10 NIX_OUTPUT_QML=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/qml NIX_OUTPUT_PLUGIN=/Users/Nasy/.nix/store/1fpr4ljk0cjybcwyy12ids5z3xf47150-qtwebengine-5.12.10-bin/lib/qt-5.12.10/plugins CONFIG+=release ) && make -f Makefile.core_generator
/Users/Nasy/.nix/store/s8hhd5ahcvff1qr6y6n9iz85kl3m7zqd-qtbase-5.12.10-dev/mkspecs/features/mac/default_post.prf:5: 'xcodeSDKInfo' is not a recognized replace function.
Project ERROR: Unknown module(s) in QT: qml-private quick-private
make[2]: *** [Makefile:68: sub-core_generator-pro-make_first] Error 3
make[2]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src/core'
make[1]: *** [Makefile:84: sub-core-make_first] Error 2
make[1]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.12.10.drv-0/qtwebengine-everywhere-src-5.12.10/src'
make: *** [Makefile:52: sub-src-make_first] Error 2
error: builder for '/Users/Nasy/.nix/store/jgkx6xqf9kiqk3n2syhy2vaivw5r3jsb-qtwebengine-5.12.10.drv' failed with exit code 2
```

--- a/pkgs/development/libraries/qt-5/modules/qtwebengine.nix
+++ b/pkgs/development/libraries/qt-5/modules/qtwebengine.nix
@@ -90,11 +90,7 @@ qtModule {
     '' else ''
     substituteInPlace src/3rdparty/chromium/base/mac/mach_port_broker.mm \
       --replace "audit_token_to_pid(msg.trailer.msgh_audit)" "msg.trailer.msgh_audit.val[5]"
-    '')
-    + ''
-    substituteInPlace src/3rdparty/chromium/sandbox/mac/BUILD.gn \
-      --replace 'libs = [ "sandbox" ]' 'libs = [ "/usr/lib/libsandbox.1.dylib" ]'
-    '');
+    ''));
 
   NIX_CFLAGS_COMPILE = lib.optionals stdenv.cc.isGNU [
     # with gcc8, -Wclass-memaccess became part of -Wall and this exceeds the logging limit
@@ -190,6 +186,7 @@ qtModule {
 
   buildInputs = optionals stdenv.isDarwin (with darwin; [
     cups
+    apple_sdk.libs.sandbox
 
     # `sw_vers` is used by `src/3rdparty/chromium/build/config/mac/sdk_info.py`
     # to get some information about the host platform.
@@ -205,14 +202,7 @@ qtModule {
         shift
       done
     '')
-
-    # For sandbox.h include
-    (runCommand "MacOS_SDK_sandbox.h" {} ''
-      install -Dm444 "${lib.getDev darwin.apple_sdk.sdk}"/include/sandbox.h "$out"/include/sandbox.h
-    '')
   ]);
-
-  __impureHostDeps = optional stdenv.isDarwin "/usr/lib/libsandbox.1.dylib";
 
   dontUseNinjaBuild = true;
   dontUseNinjaInstall = true;

--- a/pkgs/os-specific/darwin/apple-sdk/default.nix
+++ b/pkgs/os-specific/darwin/apple-sdk/default.nix
@@ -241,6 +241,18 @@ in rec {
         popd >/dev/null
       '';
     };
+
+    sandbox = stdenv.mkDerivation {
+      name = "apple-lib-sandbox";
+      dontUnpack = true;
+
+      installPhase = ''
+        mkdir -p $out/include $out/lib
+        ln -s "${lib.getDev sdk}/include/sandbox.h" $out/include/sandbox.h
+        cp "${darwin-stubs}/usr/lib/libsandbox.1.tbd" $out/lib
+        ln -s libsandbox.1.tbd $out/lib/libsandbox.tbd
+      '';
+    };
   };
 
   overrides = super: {

